apptools/undo/traitsui/trait_set_command.py

```diff
@@ -0,0 +1,82 @@
+#
+# (C) Copyright 2015 Enthought, Inc., Austin, TX
+# All right reserved.
+#
+# This file is open source software distributed according to the terms in
+# LICENSE.txt
+#
+"""
+TraitSetCommand
+===============
+
+This module provides a simple command which sets a trait value on ``do`` and
+``redo``, resets it to the previous value on ``undo``, and merges multiple
+commands on the same attribute on the same object.
+
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+# Enthought library imports
+from traits.api import Any, Str
+
+# Local library imports
+from apptools.undo.api import AbstractCommand
+
+
+class TraitSetCommand(AbstractCommand):
+    """ Command subclass which simply sets a trait on an object
+
+    This is somewhat simplistic: it assumes that nothing else is going to
+    hold on to and mutate the old value unexpectedly, which may break down
+    with things like NumPy arrays.  The main purposes is to generate commands
+    from the UndoHandler, so we can be reasonably comfortable that the values
+    are being handled sanely.
+
+    """
+
+    #-------------------------------------------------------------------------
+    # 'TraitSetCommand' interface
+    #-------------------------------------------------------------------------
+
+    #: The name of the data trait we are acting on.
+    trait_name = Str
+
+    #: The new value that the trait is being set to
+    value = Any
+
+    #: The old value of the trait
+    _old_value = Any
+
+    #-------------------------------------------------------------------------
+    # 'ICommand' interface
+    #-------------------------------------------------------------------------
+
+    #: The user-visible name of the command
+    name = Str
+
+    def do(self):
+        # we assume that copy or deep copy isn't required
+        self._old_value = getattr(self.data, self.trait_name)
+        self.redo()
+
+    def redo(self):
+        setattr(self.data, self.trait_name, self.value)
+
+    def undo(self):
+        setattr(self.data, self.trait_name, self._old_value)
+
+    def merge(self, other):
+        if (isinstance(other, self.__class__) and self.data is other.data and
+                self.trait_name == other.trait_name):
+            self.value = other.value
+            return True
```

Comment:
First question, then, is whether this whole approach of intercepting at the Handler level is worth pursuing further, in your opinion.  There are other approaches to solving this problem (such as Tony's branch above) which we could pursue for client projects, but they will require some effort, so I don't particularly want to take this further if you feel strongly that this is architecturally wrong.

I am already running into some cases where some more complex editors bypass the `Handler` completely when setting certain things (eg. selections) and so need separate shim code to capture the changes, but I think that is more a failing of the editor than an issue with using handlers, since such out-of-band changes defeat much of the point of custom handlers.

In any case, if you are of the opinion that the handler approach is worth pursing, I will implement something that allows editors, items and traits to express their mergeability, possibly with cool-down timers depending on my time constraints.  Although actual modifications to TraitsUI Editors and Items will have to wait - for the immediate needs of consulting project code trait metadata or ad-hoc Editor subclasses are probably enough.